crates/subspace-farmer-components/src/proving.rs

```diff
@@ -0,0 +1,443 @@
+    proof_of_space: PosProof,
+}
+
+struct SolutionCandidatesIterator<'a, RewardAddress, PosTable> {
+    public_key: &'a PublicKey,
+    reward_address: &'a RewardAddress,
+    sector_index: SectorIndex,
+    sector_id: SectorId,
+    s_bucket: SBucket,
+    sector_metadata: &'a SectorMetadata,
+    s_bucket_offsets: Box<[u32; Record::NUM_S_BUCKETS]>,
+    kzg: &'a Kzg,
+    erasure_coding: &'a ErasureCoding,
+    sector_contents_map: SectorContentsMap,
+    sector: Vec<u8>,
+    winning_chunks: VecDeque<WinningChunk>,
+    count: usize,
+    chunk_cache: Option<ChunkCache>,
+    _pos_table: PhantomData<PosTable>,
+}
+
+// TODO: This can be potentially parallelized with rayon
+impl<RewardAddress, PosTable> Iterator for SolutionCandidatesIterator<'_, RewardAddress, PosTable>
+where
+    RewardAddress: Copy,
+    PosTable: Table,
+{
+    type Item = Result<Solution<PublicKey, RewardAddress>, ProvingError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (chunk_offset, piece_offset, audit_chunk_offset) = {
+            let winning_chunk = self.winning_chunks.front_mut()?;
+
+            let audit_chunk_offset = winning_chunk.audit_chunk_offsets.pop_front()?;
+            let chunk_offset = winning_chunk.chunk_offset;
+            let piece_offset = winning_chunk.piece_offset;
+
+            if winning_chunk.audit_chunk_offsets.is_empty() {
+                // When all audit chunk offsets are removed, the winning chunks entry itself can be removed
+                self.winning_chunks.pop_front();
+            }
+
+            (chunk_offset, piece_offset, audit_chunk_offset)
+        };
+
+        self.count -= 1;
+
+        let chunk_cache = 'outer: {
+            if let Some(chunk_cache) = &self.chunk_cache {
+                if chunk_cache.chunk_offset == chunk_offset {
+                    break 'outer chunk_cache;
+                }
+            }
+
+            // TODO: Can potentially happen in parallel with other work below, saving a bit of
+            //  end-to-end latency
+            // Derive PoSpace table
+            let pos_table = PosTable::generate(
+                &self
+                    .sector_id
+                    .evaluation_seed(piece_offset, self.sector_metadata.history_size),
+            );
+
+            let maybe_chunk_cache: Result<_, ProvingError> = try {
+                let sector_record_chunks = read_sector_record_chunks(
+                    piece_offset,
+                    self.sector_metadata.pieces_in_sector,
+                    &self.s_bucket_offsets,
+                    &self.sector_contents_map,
+                    &pos_table,
+                    &self.sector,
+                )?;
+
+                let chunk = sector_record_chunks
+                    .get(usize::from(self.s_bucket))
+                    .expect("Within s-bucket range; qed")
+                    .expect("Winning chunk was plotted; qed");
+
+                let extended_chunks = recover_extended_record_chunks(
```

Comment:
`recover_source_record_chunks` is essentially the same as `recover_extended_record_chunks` with extension chunks removed. It has changed in https://github.com/subspace/subspace/pull/1417 to avoid extra allocation in some cases, here I have pushed a version that does in-place copies as well for efficiency purposes.